File: app/src/main/java/pl/rpieja/flat/SplashActivity.java
package pl.rpieja.flat;

import android.content.Intent;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;

import pl.rpieja.flat.authentication.AccountService;
import pl.rpieja.flat.authentication.FlatCookieJar;
import pl.rpieja.flat.api.FlatAPI;
import pl.rpieja.flat.tasks.AsyncValidate;

public class SplashActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        FlatAPI flatAPI = new FlatAPI(new FlatCookieJar(this));
        AsyncValidate.run(flatAPI, new AsyncValidate.Callable<Boolean>() {
            @Override
            public void onCall(Boolean result) {
                if (result) {
                    Intent intent = new Intent(SplashActivity.this, ChargesActivity.class);
                    startActivity(intent);
                    finish();
                } else {
                    AccountService.removeCurrentAccount(getApplicationContext());
                    Intent intent = new Intent(SplashActivity.this, LoginActivity.class);
                    startActivity(intent);
                    finish();
                }

            }
        });
    }
}

File: app/src/main/java/pl/rpieja/flat/activity/SplashActivity.kt
package pl.rpieja.flat.activity

import android.content.Intent
import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import pl.rpieja.flat.api.FlatAPI

class SplashActivity : AppCompatActivity() {

    private var validateSessionDisposable: Disposable? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        validateSessionDisposable = FlatAPI.getFlatApi(this).validateSession()
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe { validSession ->
                    val nextActivity =
                            if (validSession) MainActivity::class.java
                            else LoginActivity::class.java
                    val intent = Intent(this, nextActivity)
                    startActivity(intent)
                    finish()
                }
    }

    override fun onDestroy() {
        super.onDestroy()
        validateSessionDisposable?.dispose()
    }
}
